### PR TITLE
T5 migration to new masking interface

### PR DESCRIPTION
**Pull request by Aravind-11**
# What does this PR do?

This PR migrates the T5 model to use the new masking utilities (`masking_utils.py`) for attention mask creation.  

Fixes # (40743)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

 @vasqu @Rocketknight1 

Passed all existing test cases except for test_small_integration_test which needs GPU. I need guidance on additional test cases to be added for this. Thank you.

### CODE REVIEW

File: src/transformers/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -745,32 +747,35 @@ def forward(
...
         if self.config.is_decoder:
-            causal_mask = self._update_causal_mask(
-                attention_mask,
-                inputs_embeds,
-                cache_position,
-                past_key_values.self_attention_cache
-                if isinstance(past_key_values, EncoderDecoderCache)
-                else past_key_values,
-                output_attentions,
+            # Decoder: causal + padding combined
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py]: We don't need these comments, the function name is already pretty self-explanatory (same for below)

File: src/transformers/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -745,32 +747,35 @@ def forward(
...
-                    encoder_hidden_shape, device=inputs_embeds.device, dtype=torch.long
-                )
-            encoder_extended_attention_mask = self.invert_attention_mask(encoder_attention_mask)
+            encoder_extended_attention_mask = create_bidirectional_mask(
+                config=self.config,
+                input_embeds=encoder_hidden_states,
+                attention_mask=encoder_attention_mask,
+                cache_position=None,
+                past_key_values=None,
+            )
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py:757]: See for correct usage: https://github.com/huggingface/transformers/blob/e46c2ff32ef2c370b9214befacb0802e7ff97203/src/transformers/models/bart/modeling_bart.py#L834-L839

File: src/transformers/models/t5/modeling_t5.py [hunk 3]
```diff
@@ -745,32 +747,35 @@ def forward(
...
-            causal_mask = (1.0 - causal_mask) * torch.finfo(inputs_embeds.dtype).min
         else:
-            causal_mask = None
+            # Encoder: bidirectional attention mask
+            causal_mask = create_bidirectional_mask(
+                config=self.config,
+                input_embeds=inputs_embeds,
+                attention_mask=attention_mask,
+                cache_position=None,
+                past_key_values=None,
```
Comment (vasqu) [inline src/transformers/models/t5/modeling_t5.py]: cache_position and past_key_values are not valid fields in the function.

**Review by vasqu — COMMENTED** [review]
`create_bidirectional_mask` is not correctly used atm + let's remove unnecessary comments (including the old ones that already exist, e.g. on the cross attn mask)

I expect that some other models depend on T5, can you check with `make fix-copies`?

File: src/transformers/models/t5/modeling_t5.py (hunk 3 above)
Comment (Aravind-11) [reply src/transformers/models/t5/modeling_t5.py]: Got it. Thank you.

**Comment by Aravind-11** [comment]
> `create_bidirectional_mask` is not correctly used atm + let's remove unnecessary comments (including the old ones that already exist, e.g. on the cross attn mask)
> 
> I expect that some other models depend on T5, can you check with `make fix-copies`?

Thank you for your review @vasqu . I Changed the parameters of create_bidirectional_mask.  I ran make fix-copies and there was one dependent model mT5 that updated automatically.

The corresponding functions in modeling_mt5.py now reflect the new create_causal_mask / create_bidirectional_mask usage.

**Comment by vasqu** [comment]
run-slow: mt5, t5

**Comment by vasqu** [comment]
run-slow: mt5, t5

**Comment by vasqu** [comment]
@Aravind-11 this is a bit more complicated than anticipated, you can see the failing tests here https://github.com/huggingface/transformers/actions/runs/18776158789/job/53571258876

This means that torch export is failing. Also we would need to remove traces of the `_update_causal...` function

**Comment by vasqu** [comment]
run-slow: mt5, t5

**Comment by vasqu** [comment]
Not sure why executorch is failing with the new API tbh, reverted these parts. If you find anything that fixes this, then go ahead! I would wait a bit for you, but if the current state is ok with you, I would merge this.

cc @Cyrilvallez if you have time to check why exchanging the causal mask fails here for executorch

**Comment by Aravind-11** [comment]
> Not sure why executorch is failing with the new API tbh, reverted these parts. If you find anything that fixes this, then go ahead! I would wait a bit for you, but if the current state is ok with you, I would merge this.
> 
> cc @Cyrilvallez if you have time to check why exchanging the causal mask fails here for executorch

Thanks @vasqu! I agree, it makes sense to merge this as-is for now. The main T5 migration seems to work well and I think Executorch can be debugged later. Appreciate the review and all the follow-up commits 🙏.

**Comment by vasqu** [comment]
Checking some stuff with masking in #41852. Maybe I can fix the issue here as well then

**Comment by Aravind-11** [comment]
> Checking some stuff with masking in #41852. Maybe I can fix the issue here as well then

The changes made in that pr should be reflected here too?

**Comment by vasqu** [comment]
> Not sure why executorch is failing with the new API tbh, reverted these parts.

I meant these, working on a masking version that requires less special treatment for executorch

**Comment by Aravind-11** [comment]
> > Not sure why executorch is failing with the new API tbh, reverted these parts.
> 
> I meant these, working on a masking version that requires less special treatment for executorch

Got it.

**Comment by Aravind-11** [comment]
> > > Not sure why executorch is failing with the new API tbh, reverted these parts.
> > 
> > 
> > I meant these, working on a masking version that requires less special treatment for executorch
> 
> Got it.

Do you want me to migrate those changes here as well?

**Comment by vasqu** [comment]
No not yet at least

**Comment by vasqu** [comment]
I checked that #41852 works with t5 and all new masks API. Let's wait for that PR and then change it here to use the new mask API for causal masks as well

**Comment by Aravind-11** [comment]
> I checked that #41852 works with t5 and all new masks API. Let's wait for that PR and then change it here to use the new mask API for causal masks as well

Got it. Thank you.

**Comment by Aravind-11** [comment]
> I checked that #41852 works with t5 and all new masks API. Let's wait for that PR and then change it here to use the new mask API for causal masks as well

I think that #28818 issue also can be solved after the #41852 is approved ?

File: src/transformers/models/t5/modeling_t5.py [hunk 4]
```diff
@@ -753,25 +754,21 @@ def forward(
                 else past_key_values,
                 output_attentions,
             )
-        elif attention_mask is not None:
-            causal_mask = attention_mask[:, None, None, :]
-            causal_mask = causal_mask.to(dtype=inputs_embeds.dtype)
-            causal_mask = (1.0 - causal_mask) * torch.finfo(inputs_embeds.dtype).min
         else:
-            causal_mask = None
+            causal_mask = create_bidirectional_mask(
```
Comment (Cyrilvallez) [inline src/transformers/models/t5/modeling_t5.py]: Big mismatch between variable name and what it contains here 😉 either we're causal, or bidirectionnal

Comment (vasqu) [reply src/transformers/models/t5/modeling_t5.py]: Fair enough, should be more general (this issue was inherited from the previous code tho)

**Comment by vasqu** [comment]
run-slow: mt5, t5

**Review by vasqu — APPROVED** [review]
I pushed the new changes for causal, thx a lot for your patience @Aravind-11 

@Cyrilvallez I'd merge this tomorrow or so except you have found anything critical

**Comment by Aravind-11** [comment]
> I pushed the new changes for causal, thx a lot for your patience @Aravind-11
> 
> @Cyrilvallez I'd merge this tomorrow or so except you have found anything critical

Awesome!!

**Comment by vasqu** [comment]
Thx for sticking through this PR! Super neat to have this

**Comment by Aravind-11** [comment]
> Thx for sticking through this PR! Super neat to have this

Thank you for the code review!!
